### PR TITLE
Storage Objects Overlay Improvement

### PR DESCRIPTION
**Pull request by Robustin**
Still stretching my legs at this overlay stuff/icon stuff.

Storage objects would call update_icon whenever an item was added or removed despite 95% of them having icons that don't give a shit about their contents - so its no longer the default.

The 5% now call icon_update only when appropriate and the belts with overlays (aka what started all this) will only add or remove the overlay specific to what was added/removed and use a typecache to determine if the item even has an overlay. This is far more efficient than the current approach where the belt removes and re-adds an overlay for every item when you put in a cable coil that doesn't even have an overlay.

### CODE REVIEW

**Comment by praisenarsie** [comment]
stop this man

**Comment by Robustin** [comment]
Since I am not a coder by education/training/profession its far easier for me to work off existing code. It's still very difficult for me to do "novel" code. These performance updates feel less like skill and more like raw effort in all the reading and testing required.

**Comment by kevinz000** [comment]
kindly dont' touch storage before my pr is merged thank you i do not need more pain reeeee

**Comment by Robustin** [comment]
I thought that was merged... I don't have any idea of how much these two interact since its mostly update_icon stuff.

File: code/game/objects/items/storage/belt.dm [hunk 1]
```diff
@@ -9,23 +9,39 @@
...
 
-/obj/item/storage/belt/update_icon()
-	cut_overlays()
-	if(content_overlays)
-		for(var/obj/item/I in contents)
-			var/mutable_appearance/M = I.get_belt_overlay()
+
+/obj/item/storage/belt/remove_from_storage(obj/item/I, atom/new_location)
+	. = ..()
+	if(. && is_type_in_typecache(I, content_overlays))
```
Comment (Cyberboss) [inline code/game/objects/items/storage/belt.dm]: For here and future reference, if the first parameter of `is_type_in_typecache` is for sure a datum, the faster way to do the check is `typecache[object.type]`

File: code/game/objects/items/storage/belt.dm [hunk 2]
```diff
@@ -9,23 +9,39 @@
...
+		update_icon(I, FALSE)
+
+/obj/item/storage/belt/handle_item_insertion(obj/item/I, atom/new_location)
+	. = ..()
+	if(. && is_type_in_typecache(I, content_overlays))
```
Comment (Cyberboss) [inline code/game/objects/items/storage/belt.dm]: ^

File: code/game/objects/items/storage/belt.dm [hunk 3]
```diff
@@ -9,23 +9,39 @@
 	slot_flags = SLOT_BELT
 	attack_verb = list("whipped", "lashed", "disciplined")
 	max_integrity = 300
-	var/content_overlays = FALSE //If this is true, the belt will gain overlays based on what it's holding
+	var/list/content_overlays = list() //If this is true, the belt will gain overlays based on what it's holding
```
Comment (Cyberboss) [inline code/game/objects/items/storage/belt.dm]: remove ` = list()`

File: code/game/objects/items/storage/belt.dm [hunk 4]
```diff
@@ -9,23 +9,39 @@
...
+		if(added)
 			add_overlay(M)
-	..()
+		else
+			cut_overlay(M)
 
 /obj/item/storage/belt/Initialize()
 	. = ..()
-	update_icon()
+	if(content_overlays.len)
```
Comment (Cyberboss) [inline code/game/objects/items/storage/belt.dm]: remove `.len`

File: code/game/objects/items/storage/belt.dm [hunk 5]
```diff
@@ -9,23 +9,39 @@
...
+		content_overlays = typecacheof(content_overlays)
+		for(var/I in contents)
```
Comment (Cyberboss) [inline code/game/objects/items/storage/belt.dm]: `in src`

File: code/game/objects/items/storage/belt.dm [hunk 6]
```diff
@@ -9,23 +9,39 @@
...
+			if(is_type_in_typecache(I, content_overlays))
```
Comment (Cyberboss) [inline code/game/objects/items/storage/belt.dm]: See above

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Pull request closed without merging**